src/hotkey/hotkey_item.hpp

```diff
@@ -1,164 +1,351 @@
+
+		/**
+		 * Checks whether the hotkey bindings and scope are equal.
+		 * @param other the hokey bindings to compare against.
+		 * @return true if %other has same scope and bindings.
+		 */
+		virtual bool bindings_equal(hotkey_ptr other);
+
+		virtual ~hotkey_base() {}
+
+		protected:
+		/**
+		 *  This is invoked by hotkey_base::get_name and must be implemented by subclasses.
+		 *  Keyboard modifiers are handled in this class, other hotkeys in the respective classes
+		 */
+		virtual const std::string get_name_helper() const = 0;
+		/**
+		 * This is invoked by hotkey_base::matches as a helper for the concrete classes.
+		 * Implementing classes should only check their parts of the hotkey.
+		 * @param event The SDL_Event being generated.
+		 * @returns true if they match, false otherwise.
+		 */
+		virtual bool matches_helper(const SDL_Event &event) const = 0;
+		virtual void save_helper(config& cfg) const = 0;
+		/**
+		 * This is invoked by hotkey_base::bindings_equal as a helper for the concrete classes.
+		 * Implementing classes should only check their parts of the hotkey.
+		 * @param other The other hotkey the check against. Not guaranteed to be the same subclass.
+		 * @returns true if they match, false otherwise.
+		 */
+		virtual bool bindings_equal_helper(hotkey_ptr other) const = 0;
+
+		/**
+		 * The command that should be executed, or "null".
+		 */
+		std::string command_;
+
+		/**
+		 *
+		 */
+		bool is_default_;
+		/*
+		 * Keyboard modifiers. Treat as opaque, only do comparisons.
+		 *
+		 */
+		unsigned int mod_;
+	};
+
+	/**
+	 * This class is responsible for handling keys, not modifiers.
+	 */
+	class hotkey_keyboard: public hotkey_base {
+		public:
+		/**
+		 * Initialise new instance of this class that has no key associated with is.
+		 */
+		hotkey_keyboard() : hotkey_base(), scancode_(SDL_SCANCODE_UNKNOWN) {}
+
+		/**
+		 * Set the scancode associated with this class.
+		 * @param scancode The SDL_Scancode that this hotkey should be associated with
+		 */
+		void set_scancode(SDL_Scancode scancode) {scancode_ = scancode;}
+
+		/**
+		 * Checks whether this hotkey has been set to a sensible value.
+		 * @ return true if it is a known key
+		 */
+		virtual bool valid() const {return scancode_ != SDL_SCANCODE_UNKNOWN;}
+
+		protected:
+		SDL_Scancode scancode_;
+
+		virtual void save_helper(config& cfg) const;
+		virtual const std::string get_name_helper() const;
+		virtual bool matches_helper(const SDL_Event &event) const;
+		virtual bool bindings_equal_helper (hotkey_ptr other) const;
+	};
+
+	/**
+	 * This class is used to return non-valid results in order to save
+	 * other people from null checks.
+	 */
+	class hotkey_void: public hotkey_base {
+		public:
+		hotkey_void() : hotkey_base() {}
+		virtual bool valid() const {return false;}
+		protected:
+		virtual void save_helper(config& cfg) const {UNUSED(cfg);};
+		virtual const std::string get_name_helper() const {return "";}
```

Comment:
It would be better to just omit the parametername and to remove the UNUSED then.